### PR TITLE
Disable plugins loading

### PR DESCRIPTION
**Pull request by aobolensk**

### CODE REVIEW

**Comment by dmitryermilov** [comment]
@aobolensk , could you describe which scenarios did you test?

**Comment by aobolensk** [comment]
> @aobolensk , could you list scenarios which you tested? Just to be on the same page, all plug-in *.so files and plugin.cfg file were removed during these test, right?

Yes, I have removed all plugin *.so (I checked path provided to LoadPlugin function).
Scenarios:
```
./sample_encode vp9 -i input.yuv -o out.vp9 -w 1920 -h 1080 -p vp9e_hw
./sample_encode h265 -i input.yuv -o out.h265 -w 1920 -h 1080 -p hevce_hw
```

**Comment by dmitryermilov** [comment]
> Yes, I have removed all plugin *.so (I checked path provided to LoadPlugin function).
what about plugins.cfg file ?

**Comment by aobolensk** [comment]
Double checked with removed plugins.cfg

**Review by dvrogozh — COMMENTED** [review]
What's the purpose of this draft? Effectively this looks to be backward incompatible change for current version of msdk.

**Comment by dmitryermilov** [comment]
> Effectively this looks to be backward incompatible change for current version of msdk.

Why? It's compatible.

**Comment by dvrogozh** [comment]
No. You disable plugins load, i.e. applications which currently load plugins such as HEVC encoder/decoder will break if you merge in this change. First candidate to break is ffmpeg. They definitely have plugin load path, but I am not sure whether they implemented this to be optional. At some point this was not optional.

**Comment by dmitryermilov** [comment]
> No. You disable plugins load, i.e. applications which currently load plugins such as HEVC encoder/decoder will break if you merge in this change.

You're repeating. Again, why do you think so? All plug-ins except LA are part of library for > 1 year.

**Comment by dvrogozh** [comment]
And you don't answer the question:): what this change is about and what are you ultimately trying to achieve?

**Comment by dvrogozh** [comment]
Actually, I am not against the change (I somewhat overlooked that you return MFX_ERR_NONE from plugin functions). But I want more clear justification of what you are doing and why in the commit message. I suggest the following:
1. First commit: disable load of msdk plugins which actually correspond to codecs implemented inside the library (i.e. return MFX_ERR_NONE from plugin load/unload functions pretending that we loaded everything)
2. Second commit: remove msdk plugins from the build and msdk distribution
Above applies to all msdk plugins except LA since it's the only real plugin which remains.

**Comment by dmitryermilov** [comment]
@xhaihao , this PR is partially about what we discussed today. JFYI.

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 1]
```diff
@@ -472,6 +482,16 @@ mfxStatus MFXVideoUSER_Load(mfxSession session, const mfxPluginUID *uid, mfxU32
 
 mfxStatus MFXVideoUSER_LoadByPath(mfxSession session, const mfxPluginUID *uid, mfxU32 version, const mfxChar *path, mfxU32 /*len*/)
 {
+  if (*uid == MFX_PLUGINID_HEVCD_SW ||
```
Comment (dvrogozh) [inline api/mfx_dispatch/linux/mfxloader.cpp:485]: create a (static inline) function which makes this check...

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 2]
```diff
@@ -324,6 +324,9 @@ mfxStatus PluginCtx::Unload()
 
 mfxStatus LoaderCtx::LoadPlugin(const mfxPluginUID &uid, mfxU32 version, const char *path)
 {
+  if (uid != MFX_PLUGINID_H264LA_HW) {
```
[the remark this answers is not in the document]
Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Done

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 3]
```diff
@@ -472,6 +478,9 @@ mfxStatus MFXVideoUSER_Load(mfxSession session, const mfxPluginUID *uid, mfxU32
 
 mfxStatus MFXVideoUSER_LoadByPath(mfxSession session, const mfxPluginUID *uid, mfxU32 version, const mfxChar *path, mfxU32 /*len*/)
 {
+  if (*uid != MFX_PLUGINID_H264LA_HW) {
```
[the remark this answers is not in the document]
Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Done

File: api/mfx_dispatch/linux/mfxloader.cpp (hunk 1 above)
Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Done

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 4]
```diff
@@ -429,8 +429,24 @@ mfxStatus MFXClose(mfxSession session)
   }
 }
 
+static inline bool IsDisabledMsdkPlugin(const mfxPluginUID *uid)
```
[the remark this answers is not in the document]
Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Done

File: api/mfx_dispatch/linux/mfxloader.cpp [hunk 5]
```diff
@@ -429,8 +429,24 @@ mfxStatus MFXClose(mfxSession session)
...
+    *uid == MFX_PLUGINID_HEVCE_HW ||
+    *uid == MFX_PLUGINID_VP8D_HW ||
+    *uid == MFX_PLUGINID_VP8E_HW ||
+    *uid == MFX_PLUGINID_VP9D_HW ||
+    *uid == MFX_PLUGINID_VP9E_HW);
+}
+
 mfxStatus MFXVideoUSER_Load(mfxSession session, const mfxPluginUID *uid, mfxU32 version)
 {
+  if (IsDisabledMsdkPlugin(uid)) {
```
[the remark this answers is not in the document]
Comment (aobolensk) [reply api/mfx_dispatch/linux/mfxloader.cpp]: Done

**Comment by xhaihao** [comment]
My testing shows ffmpeg-qsv and gst-msdk work with this change, but I didn't do a full cycle testing.

**Comment by dvrogozh** [comment]
Let's go with 2 separate PRs. First, disable plugins in dispatcher and continue discussion of physical removal of plugins in a separate PR...

**Comment by aobolensk** [comment]
> Let's go with 2 separate PRs. First, disable plugins in dispatcher and continue discussion of physical removal of plugins in a separate PR...

Created new PR: https://github.com/Intel-Media-SDK/MediaSDK/pull/2490

**Review by dvrogozh — APPROVED** [review]

**Review by onabiull — APPROVED** [review]

**Pull request merged**